File: iosApp/iosApp/IapBridge.swift
import Cmp
import Foundation
import StoreKit

/// Swift implementation of the Kotlin `IapBridge` interface (defined in `:shared` commonMain).
/// 호출자(`IosIapClient`)가 callback 을 suspend 함수로 변환한다.
///
/// StoreKit2 (iOS 15+) — `Product.purchase()` 의 시스템 결제 popup 은 Apple 이 그린다.
/// 본 클래스는 trigger + 결과 매핑만; UI 모방 없음 (App Store 가이드라인 3.1.1 / 4.1 / 4.5).
///
/// 세션 동안 `Product` 객체는 캐시 — 매 구매마다 카탈로그 RTT (~100~500ms) 방지.
///
/// **출시 전 필수**:
///   - App Store Connect 에 productId 등록 (consumable IAP).
///   - Xcode > Capabilities > In-App Purchase 추가.
///   - Sandbox tester 계정으로 결제 흐름 verify.
///   - `Transaction.updates` 장기 listener 추가 — Ask-to-Buy 승인 / Family Sharing / crash
///     직후 복구 / 환불 revocation 처럼 `Product.purchase()` 호출 밖에서 도착하는 transaction
///     도 처리해야 결제 후 credit 누락 안 됨. (현 v1 은 명시적 구매 흐름만 커버)
final class IapBridgeImpl: NSObject, IapBridge {

    // outcome 문자열은 Kotlin `IapBridge` 의 expected 값과 동기. 변경 시 양쪽 모두 갱신.
    private enum Outcome {
        static let success = "success"
        static let cancelled = "cancelled"
        static let failed = "failed"
    }

    private var productCache: [String: Product] = [:]
    private let cacheQueue = DispatchQueue(label: "vibi.iap.cache")

    func purchase(
        productId: String,
        callback: @escaping (String, String?, String?, String?) -> Void
    ) -> IapCancellable {
        let task = Task {
            do {
                let product = try await cachedProduct(productId: productId)
                guard let product = product else {
                    callback(Outcome.failed, nil, nil, "product_not_found")
                    return
                }
                try Task.checkCancellation()
                let result = try await product.purchase()
                try Task.checkCancellation()

                switch result {
                case .success(let verification):
                    switch verification {
                    case .verified(let tx):
                        // 영수증 — Transaction.jsonRepresentation 은 Apple 서명된 JWS.
                        // BFF 가 App Store Server API 로 재검증하므로 그대로 base64 인코딩 전달.
                        let receiptB64 = tx.jsonRepresentation.base64EncodedString()
                        // 소비성 (consumable) — finish 호출해야 다음 구매 가능. BFF 검증 후 호출이
                        // 정석이지만 v1 stub 영수증 검증 단계라 여기서 즉시 finish.
                        await tx.finish()
                        callback(Outcome.success, "\(tx.id)", receiptB64, nil)
                    case .unverified(_, let error):
                        callback(Outcome.failed, nil, nil, "verification_failed: \(error.localizedDescription)")
                    }
                case .userCancelled:
                    callback(Outcome.cancelled, nil, nil, nil)
                case .pending:
                    // Ask-to-Buy / SCA 보류. 결과는 Transaction.updates 로 별도 도착 (출시 전 TODO).
                    callback(Outcome.failed, nil, nil, "purchase_pending")
                @unknown default:
                    callback(Outcome.failed, nil, nil, "unknown_result")
                }
            } catch is CancellationError {
                // 호출자가 취소 — callback 호출 안 함 (Kotlin 측 continuation 도 이미 cancel 상태).
                return
            } catch {
                callback(Outcome.failed, nil, nil, error.localizedDescription)
            }
        }
        return CancellableTask(task: task)
    }

    func restorePurchases(callback: @escaping (String, String?) -> Void) -> IapCancellable {
        let task = Task {
            do {
                // consumable 은 보통 복원 대상 아님. 권장사항 충족 + 향후 비소비성/구독 도입 시
                // 즉시 동작하도록 AppStore.sync() 호출.
                try await AppStore.sync()
                try Task.checkCancellation()
                callback(Outcome.success, nil)
            } catch is CancellationError {
                return
            } catch {
                callback(Outcome.failed, error.localizedDescription)
            }
        }
        return CancellableTask(task: task)
    }

    private func cachedProduct(productId: String) async throws -> Product? {
        if let cached = cacheQueue.sync(execute: { productCache[productId] }) {
            return cached
        }
        let products = try await Product.products(for: [productId])
        guard let product = products.first else { return nil }
        cacheQueue.sync { productCache[productId] = product }
        return product
    }
}

/// Kotlin `IapCancellable` 의 Swift 구현. `IosIapClient.suspendCancellableCoroutine.invokeOnCancellation`
/// 에서 호출되어 Swift Task 까지 취소 — 결제 popup 중 sheet dismiss 시 receipt 누수 방지.
private final class CancellableTask: IapCancellable {
    private let task: Task<Void, Never>
    init(task: Task<Void, Never>) { self.task = task }
    func cancel() { task.cancel() }
}

File: shared/src/commonMain/kotlin/com/vibi/shared/platform/IapBridge.kt
package com.vibi.shared.platform

/**
 * Swift 가 구현해 K/N 으로 주입하는 IAP bridge. iOS 만 사용 — Android 는 별도 `BillingClient`
 * 통합 시점에 같은 패턴으로 확장 가능하지만 v1 은 stub.
 *
 * StoreKit2 (`Product.purchase()` 등) 는 Swift-only API 라 Kotlin/Native cinterop 으로 직접
 * 호출할 수 없다. Apple Sign-In 과 동일한 callback bridge 방식으로 Swift 측이 결제 흐름을
 * 처리하고 결과만 Kotlin 으로 전달.
 *
 * suspend 함수를 Swift 가 구현하기 어려워 callback 으로 단순화. `IosIapClient` 가 wrapper 로
 * `suspendCancellableCoroutine` 변환 + 반환된 [IapCancellable] 로 코루틴 취소 시 Swift Task
 * 도 취소.
 *
 * outcome 문자열 단일 source — Swift 와 Kotlin 양쪽이 같은 값 사용. 변경 시 두 쪽 모두 갱신.
 * (`"success"` | `"cancelled"` | `"failed"`)
 *
 * **출시 전 TODO**: `Transaction.updates` 장기 listener 추가 — Ask-to-Buy 승인 / Family
 * Sharing / crash 직후 복구 / 환불 revocation 처럼 `Product.purchase()` 호출 밖에서
 * 도착하는 transaction 도 처리해야 사용자가 결제 후 credit 누락되지 않는다.
 */
interface IapBridge {
    /**
     * StoreKit2 `Product.purchase()` 트리거. callback 은 정확히 1회 호출.
     *
     * @return 코루틴이 취소되면 [IapCancellable.cancel] 로 Swift Task 까지 전파. 이미 결제 popup
     *   이 떠 있으면 Swift 측이 무시 (Apple 가이드).
     * @param outcome `"success"` | `"cancelled"` | `"failed"`
     * @param transactionId `Transaction.id` 문자열 — success 일 때만 non-null, BFF idempotency key.
     * @param receipt `Transaction.jsonRepresentation` 의 base64 — success 일 때만 non-null.
     * @param errorMessage failed 일 때 사유.
     */
    fun purchase(
        productId: String,
        callback: (
            outcome: String,
            transactionId: String?,
            receipt: String?,
            errorMessage: String?,
        ) -> Unit,
    ): IapCancellable

    /**
     * `AppStore.sync()` 호출. consumable (크레딧) 은 보통 복원 대상 아니지만 App Store
     * 가이드라인 권장사항 충족 + 향후 비소비성/구독 도입 시 즉시 동작하도록 노출.
     *
     * @param outcome `"success"` | `"cancelled"` | `"failed"`
     */
    fun restorePurchases(
        callback: (outcome: String, errorMessage: String?) -> Unit,
    ): IapCancellable
}

/**
 * 진행 중인 StoreKit Task 의 취소 핸들. `suspendCancellableCoroutine.invokeOnCancellation`
 * 에서 호출되어 sheet dismiss / 화면 이탈 시 Swift 측 Task 도 같이 끊는다.
 */
interface IapCancellable {
    fun cancel()
}
